### PR TITLE
Take seed rather than Key for some PRNG algorithms

### PR DESCRIPTION
**Pull request by sr5434**
For some random number generation algorithms(randint, shuffle, choice, permutation, normal, multivariate_normal, truncated_normal, bernoulli), instead of requiring a PRNGKey, the user can pass a seed into the algorithm. This means that the user doesn't have to create a random key when using these algorithms.

### CODE REVIEW

**Comment by hawkinsp** [comment]
I suspect we are not going to accept this PR because the use of a `key` is quite intentional here, but I'll let @froystig who is the PRNG expert comment.

Note that these are not stateful PRNGs, so "seed" would be misleading at best.

**Comment by sr5434** [comment]
But we are passing the seed into the Key, right? Wouldn't it be easier to use if we passed the seed for the key generator into the PRNG, and took care of generating the key for the user?

**Comment by mattjj** [comment]
Thanks for the idea, but I agree with what Peter wrote. Some additional points:
1. Passing around ints means the program expresses strictly less information than if it were to pass around PRNGKeys instead. Concretely that would defeat several of our ideas for PRNGKey reuse checking (which would check reuse of PRNGKeys but not ints, so the more PRNGKey-typed values in the program the safer we can be).
2. It simultaneously makes the code less abstract and less flexible, since the PRNGKey details aren't abstracted away (and so we can't switch PRNG key generation strategies).
3. If someone wants this convenience layer today, even at the above costs, they can pretty easily just wrap all the functions. But we tend to shy away from baking tradeoff-inducing convenience wrappers into JAX itself.

**Comment by sr5434** [comment]
Couldn’t we have the current functions and wrapper functions so that users can choose which function they can use?

**Comment by mattjj** [comment]
Certainly we could, but we don't bake those convenience wrappers into JAX itself.

**Comment by sr5434** [comment]
Oh ok.

**Comment by sr5434** [comment]
> Thanks for the idea, but I agree with what Peter wrote. Some additional points:
> 
> 1. Passing around ints means the program expresses strictly less information than if it were to pass around PRNGKeys instead. Concretely that would defeat several of our ideas for PRNGKey reuse checking (which would check reuse of PRNGKeys but not ints, so the more PRNGKey-typed values in the program the safer we can be).
> 2. It simultaneously makes the code less abstract and less flexible, since the PRNGKey details aren't abstracted away (and so we can't switch PRNG key generation strategies).
> 3. If someone wants this convenience layer today, even at the above costs, they can pretty easily just wrap all the functions. But we tend to shy away from baking tradeoff-inducing convenience wrappers into JAX itself.

Can’t we trust the user to not reuse keys? Also, couldn’t we add an argument to the function call allowing them to switch key generation strategies?

**Comment by froystig** [comment]
We don't do reuse checks by default, so you could say we "trust the user" all the time, but we offer a means of opting into a checker (under development). It's helpful in debugging.

I'm going to close. Discussion can always continue of course. This is also a major breaking change to the public API, and it adds complexity in making the functions in `jax.random` inconsistent with one another, so we'd need a strong case for it.

**Comment by sr5434** [comment]
If we have wrappers alongside the original functions for all PRNG algorithms, would it be possible to merge?

**Comment by soraros** [comment]
> - If someone wants this convenience layer today, even at the above costs, they can pretty easily just wrap all the functions. But we tend to shy away from baking tradeoff-inducing convenience wrappers into JAX itself.

@sr5434, based on @mattjj's bullet point above, I'd say highly unlikely.
Personally, I think JAX's random module is very well-designed. Maybe this [tutorial](https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html) can convince you `KeyArray` is a better idea than NumPy-style random seed.

BTW, I suppose this discussion is better continued in [Github Discussions](https://github.com/google/jax/discussions).
